src/graph.h

```diff
@@ -0,0 +1,170 @@
+//
+// libsemigroups - C++ library for semigroups and monoids
+// Copyright (C) 2017 James D. Mitchell
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+#ifndef LIBSEMIGROUPS_SRC_GRAPH_H_
+#define LIBSEMIGROUPS_SRC_GRAPH_H_
+
+#include <iostream>
+#include <stack>
+#include "recvec.h"
+
+namespace libsemigroups {
+  // Represents vertices as rows
+  // There is an edge (i, j) in the graph iff j occurs in row i
+  //
+  // TODO: proper commentary (makedoc)
+  // TODO: maybe mutability
+  class Graph : public RecVec<size_t> {
+    typedef RecVec<size_t, std::allocator<size_t>> base_recvec;
+
+   public:
+    // if we have graphs with max of size_t vertices we have bigger problems
+    static const size_t UNDEFINED;
+
+    explicit Graph(size_t degree, size_t nr_vertices = 0)
+        : base_recvec(degree, nr_vertices, UNDEFINED), _cc_ids(),
+          _has_scc(false), _next_edge_pos(nr_vertices), degree(degree) {}
+
+    Graph &operator=(Graph const &graph) = delete;
+
+    void inline set(size_t i, size_t j, size_t k) {
+      base_recvec::set(i, j, k);
+      _has_scc = false;
+    }
+
+    void inline add_nodes(size_t nr) {
+      base_recvec::add_rows(nr);
+      for (size_t i = 0; i < nr; ++i) {
+        _next_edge_pos.push_back(0);
+      }
+      _has_scc = false;
+    }
+
+    size_t inline nr_nodes() { return nr_rows(); }
+
+    void inline add_edge(size_t i, size_t j) {
+      LIBSEMIGROUPS_ASSERT(_next_edge_pos[i] < degree);
+      LIBSEMIGROUPS_ASSERT(i < _nr_rows);
+      set(i, _next_edge_pos[i], j);
+      ++_next_edge_pos[i];
+    }
+
+    size_t inline nr_edges() {
```

Comment:
Can't the same be achieved by looping over `_next_edge_pos`? And summing `_nr_rows - _next_edge_pos[i]`?